feat(rpc): introduce FilterIter

ValuedMammal opened the pull request:
The PR adds a `bip158` module to the `bdk_bitcoind_rpc` crate along with a new type `FilterIter` that can be used for retrieving blocks from a full node which contain transactions relevant to a list of script pubkeys.

### Notes to the reviewers

### Changelog notice

`bdk_bitcoind_rpc`: Added `bip158` module as a means of updating `bdk_chain` structures

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

ValuedMammal committed:
feat(rpc): introduce `FilterIter`

evanlinjin commented:
This is what we've all been waiting for.

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -0,0 +1,261 @@
+                }
+            }
+            let start_height = self.blocks.keys().next().expect("blocks not empty");
+            self.height = cp.height().min(*start_height);
```
This looks to be inefficient in that it could cause us to re-emit full blocks redundantly. I think the start height should always begin one block after the point of agreement with the local chain

LLFourn commented:
ConceptACK. Don't have time for a full review now. @evanlinjin?

oleonardolima commented on crates/bitcoind_rpc/examples/bip158.rs:
```diff
@@ -0,0 +1,75 @@
+    // Assume a minimum birthday height
+    let block = BlockId {
+        height: 205_000,
+        hash: "0000002bd0f82f8c0c0f1e19128f84c938763641dba85c44bdb6aed1678d16cb".parse()?,
```
nit: I guess these fields could be moved as constants above, wdyt?

oleonardolima commented on crates/bitcoind_rpc/examples/bip158.rs:
```diff
@@ -0,0 +1,75 @@
+
+    // Configure RPC client
+    let rpc_client = bitcoincore_rpc::Client::new(
+        "127.0.0.1:38332",
```
nit: It could be used either as a constant above, or an environment variable (preferrable in my opinion).

oleonardolima commented on crates/bitcoind_rpc/tests/bip158.rs:
```diff
@@ -0,0 +1,114 @@
+            hash: bdk_core::bitcoin::hashes::Hash::hash($hash.as_bytes()),
+        }
+    }};
+}
```
nit: you can now use this `block_id` from `bdk_testenv`.

oleonardolima commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -177,16 +177,18 @@ impl<'c, C: RpcApi> Iterator for FilterIter<'c, C> {
     fn next(&mut self) -> Option<Self::Item> {
+        if self.spks.is_empty() {
+            return None;
+        }
```
I assume that the user would properly use the `extend` or `add` methods, however, I wonder how could we make it clear here that it's missing the `spks`, otherwise the user would only get Nones without much information 🤔

ValuedMammal replied:
In this case would it be better to return `Event::NoMatch`? If I recall, the reason for adding this check was that `FilterIter` seemed to be returning empty blocks as matches even without spks (on regtest) which is also confusing. I'll do some more testing though to confirm it.

oleonardolima replied:
I would say it could be a specific error variant indicating the spks are empty.

ValuedMammal commented:
Thanks for the review @oleonardolima

ValuedMammal commented:
Rebased and addressed comments from @oleonardolima

nymius commented on crates/bitcoind_rpc/tests/test_filter_iter.rs:
```diff
@@ -0,0 +1,142 @@
+        .get_new_address(None, None)?
+        .assume_checked();
+    let spk = addr.script_pubkey();
+    let _ = env.mine_blocks(9, Some(addr))?;
```
If the goal of this test is to check the behavior when no spk has been added to the filter, I would avoid reiterative checks over the same thing. In this case, it would mean to just mine 1 or 2 blocks.

```suggestion
    let _ = env.mine_blocks(1, Some(addr))?;

    let mut iter = FilterIter::new_with_height(env.rpc_client(), 1);
    assert_eq!(iter.get_tip().unwrap().map(|b| b.height), Some(2));

    // iterator should return two errors
    for res in iter.by_ref().take(2) {
        assert!(matches!(res, Err(Error::NoScripts)));
```

nymius commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -0,0 +1,277 @@
+                Ok(Some(Event::NoMatch(height)))
+            };
+
+            self.next_filter = self.next_filter()?;
```
I felt `next_iter` field was redundant, as it is a result of the state of other fields already kept by `FilterIter`, and not something accessed by the other struct methods besides `next` from `Iterator`.  If you agree, I came up with the following:

```diff
diff --git a/crates/bitcoind_rpc/src/bip158.rs b/crates/bitcoind_rpc/src/bip158.rs
index 2c5bc114..5419716b 100644
--- a/crates/bitcoind_rpc/src/bip158.rs
+++ b/crates/bitcoind_rpc/src/bip158.rs
@@ -33,8 +33,6 @@ pub struct FilterIter<'c, C> {
     cp: Option<CheckPoint>,
     // blocks map
     blocks: BTreeMap<Height, BlockHash>,
-    // next filter
-    next_filter: Option<NextFilter>,
     // best height counter
     height: Height,
     // stop height
@@ -49,7 +47,6 @@ impl<'c, C: RpcApi> FilterIter<'c, C> {
             spks: vec![],
             cp: None,
             blocks: BTreeMap::new(),
-            next_filter: None,
             height,
             stop: 0,
         }
@@ -127,9 +124,6 @@ impl<'c, C: RpcApi> FilterIter<'c, C> {
 
         self.stop = tip_height;
 
-        // get the first filter
-        self.next_filter = self.next_filter()?;
-
         Ok(Some(BlockId {
             height: tip_height,
             hash: tip_hash,
@@ -177,31 +171,27 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
     type Item = Result<Event, Error>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let (block, filter) = self.next_filter.clone()?;
-
         (|| -> Result<_, Error> {
             // if the next filter matches any of our watched spks, get the block
             // and return it, inserting relevant block ids along the way
-            let height = block.height;
-            let hash = block.hash;
-
-            let result = if self.spks.is_empty() {
-                Err(Error::NoScripts)
-            } else if filter
-                .match_any(&hash, self.spks.iter().map(|script| script.as_bytes()))
-                .map_err(Error::Bip158)?
-            {
-                let block = self.client.get_block(&hash)?;
-                self.blocks.insert(height, hash);
-                let inner = EventInner { height, block };
-                Ok(Some(Event::Block(inner)))
-            } else {
-                Ok(Some(Event::NoMatch(height)))
-            };
-
-            self.next_filter = self.next_filter()?;
-
-            result
+            self.next_filter()?.map_or(Ok(None), |(block, filter)| {
+                let height = block.height;
+                let hash = block.hash;
+
+                if self.spks.is_empty() {
+                    Err(Error::NoScripts)
+                } else if filter
+                    .match_any(&hash, self.spks.iter().map(|script| script.as_bytes()))
+                    .map_err(Error::Bip158)?
+                {
+                    let block = self.client.get_block(&hash)?;
+                    self.blocks.insert(height, hash);
+                    let inner = EventInner { height, block };
+                    Ok(Some(Event::Block(inner)))
+                } else {
+                    Ok(Some(Event::NoMatch(height)))
+                }
+            })
         })()
         .transpose()
     }
```

ValuedMammal replied:
Brilliant. will fix

ValuedMammal commented:
In the last push

- Removed unneeded `next_filter` field
- Fixed up test checking for `Error::NoScripts`
- Added test `filter_iter_returns_matched_blocks`

CI failure seems unrelated

oleonardolima reviewed (commented):
cACK db65d96a98073b1d96f62a0b5643fa997fadb4ca

ValuedMammal commented:
Rebased

nymius commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -0,0 +1,267 @@
+        self.spks.push(spk);
+    }
+
+    /// Get the next filter and increment the current best height.
```
Shouldn't this be _line comments_ (`// Get the next filt...`) instead of _line docs_ (`/// Get the next filt...`)? If not, documentation for this private method will only be shown if compiled with `cargo doc --document-private-items`.

nymius replied:
Addressed by @ValuedMammal in weekly call. Is used to have documentation snippets of private methods in text editor, and I think is reasonable to have it there, if at the end is not going to collude with the public documentation.

nymius reviewed (approved):
tACK 7ad95d35

ValuedMammal commented:
Rebased

notmandatory committed:
docs(rpc): add README and print ext address for FilterIter example

notmandatory reviewed (approved):
tACK a6364e20500fba7c16f31fa97d106893110751c6

This is cool stuff, I added some docs to make it easier for others to run.

ValuedMammal commented:
Thanks @notmandatory

evanlinjin commented:
We need to test this against reorgs. It seems like the architecture of this may result in inconsistent state of checkpoints during reorgs.

1. Have a chain with blocks `.. [100:A] [101:B]` with relevant txs in both A and B.
2. Sync up to 100 (calling `.next`).
3. Reorg so that blocks from 100 are replaced (inclusive). `.. [100:A'] [101:B']`.
4. Call `.next` again.
5. Check that checkpoints end in `.. [100:A'] [101:B']` and we have emitted relevant txs in `A'` and `B'`.
